### PR TITLE
HDDS-3608. NPE while process a pipeline report when PipelineQuery absent in query2OpenPipelines

### PR DESCRIPTION
**Pull request by maobaolong**
## What changes were proposed in this pull request?

Fix a NPE while SCM receive and handle a datanode pipeline report, PipelineStateMap will get the pipelineList from `query2OpenPipelines`, and put the `updatedPipeline` to the list, but if the list is NULL, NPE will occur. 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3608

## How was this patch tested?

No need

### CODE REVIEW

**Comment by maobaolong** [comment]
Hey @linyiqun , would you like to take a look at this PR? Thank you.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineStateMap.java [hunk 1]
```diff
@@ -374,7 +374,12 @@ Pipeline updatePipelineState(PipelineID pipelineID, PipelineState state)
     PipelineQuery query = new PipelineQuery(pipeline);
     if (updatedPipeline.getPipelineState() == PipelineState.OPEN) {
       // for transition to OPEN state add pipeline to query2OpenPipelines
-      query2OpenPipelines.get(query).add(updatedPipeline);
+      List<Pipeline> pipelineList = query2OpenPipelines.get(query);
+      if (pipelineList == null) {
+        pipelineList = new CopyOnWriteArrayList<>();
+        query2OpenPipelines.put(query, pipelineList);
+      }
+      pipelineList.add(updatedPipeline);
```
Comment (linyiqun) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineStateMap.java:382]: Besides above change, Can you also do the empty check for another condition branch?
    } else {
      // for transition from OPEN to CLOSED state remove pipeline from
      // query2OpenPipelines
      query2OpenPipelines.get(query).remove(pipeline);   <---
    }
There is also a chance that remove operation can throw NPE.

**Comment by maobaolong** [comment]
@linyiqun Thanks for your review comments, i've add an additional empty check, PTAL.

**Comment by linyiqun** [comment]
LGTM, +1. Thanks for addressing the comment, @maobaolong .
I didn't have the write access to merge the PR now. So please wait others for helping review and merging.

**Comment by maobaolong** [comment]
@linyiqun Thank you for your review, it's important to receive your +1

**Comment by maobaolong** [comment]
@elek Please take a look at this PR, thank you.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineStateMap.java [hunk 2]
```diff
@@ -372,13 +372,20 @@ Pipeline updatePipelineState(PipelineID pipelineID, PipelineState state)
     Pipeline updatedPipeline = pipelineMap.compute(pipelineID,
         (id, p) -> Pipeline.newBuilder(pipeline).setState(state).build());
     PipelineQuery query = new PipelineQuery(pipeline);
+    List<Pipeline> pipelineList = query2OpenPipelines.get(query);
     if (updatedPipeline.getPipelineState() == PipelineState.OPEN) {
       // for transition to OPEN state add pipeline to query2OpenPipelines
-      query2OpenPipelines.get(query).add(updatedPipeline);
+      if (pipelineList == null) {
```
Comment (xiaoyuyao) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineStateMap.java:378]: query2OpenPipelines has been initialized in initializeQueryMap() based on the RepType/ReFactor and has never been removed from the map. Correct me if I'm wrong, the pipelineList should never be null unless different RepType/Factor are specified in the query, which is currently impossible.

Comment (maobaolong) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineStateMap.java:378]: @xiaoyuyao Yeah, you are right while we do something initialized in initializeQueryMap(), but i am working on query unknown replication query dynamically, so initializeQueryMap() just put the known replication factor into the query2OpenPipelines. Further more, i have an idea to remove initializeQueryMap, instead, using a lazy way to add and remove items from/to the query2OpenPipelines, it is only an idea, but for this PR, it makes NPE gone away.

**Comment by maobaolong** [comment]
@xiaoyuyao Sorry, i don't find your comments inline? Could you show me where is the inline comments you mean?

**Comment by xiaoyuyao** [comment]
> @xiaoyuyao Sorry, i don't find your comments inline? Could you show me where is the inline comments you mean?

The comments is just above:

query2OpenPipelines has been initialized in initializeQueryMap() based on the RepType/ReFactor and has never been removed from the map. Correct me if I'm wrong, the pipelineList should never be null unless different RepType/Factor are specified in the query, which is currently impossible.

**Comment by maobaolong** [comment]
@xiaoyuyao Sorry, i saw your reply. You are right with the condition `query2OpenPipelines has been initialized in initializeQueryMap() based on the RepType/ReFactor and has never been removed from the map`, but for the logic of the method updatePipelineState itself, i think it is better to have a non-null check, and do something to fall back. Actually, i'm just doing something broken the condition you've mentioned before, so i met the NPE already.

**Comment by xiaoyuyao** [comment]
Sounds good to me. Thanks @maobaolong and all for the discussion.

**Pull request merged**